sql/core/src/main/scala/org/apache/spark/sql/execution/python/WindowInPandasExec.scala

```diff
@@ -144,24 +282,107 @@ case class WindowInPandasExec(
         queue.close()
       }
 
-      val inputProj = UnsafeProjection.create(allInputs, child.output)
-      val pythonInput = grouped.map { case (_, rows) =>
-        rows.map { row =>
-          queue.add(row.asInstanceOf[UnsafeRow])
-          inputProj(row)
+      val stream = iter.map { row =>
+        queue.add(row.asInstanceOf[UnsafeRow])
+        row
+      }
+
+      val pythonInput = new Iterator[Iterator[UnsafeRow]] {
+
+        // Manage the stream and the grouping.
+        var nextRow: UnsafeRow = null
+        var nextGroup: UnsafeRow = null
+        var nextRowAvailable: Boolean = false
+        private[this] def fetchNextRow() {
+          nextRowAvailable = stream.hasNext
+          if (nextRowAvailable) {
+            nextRow = stream.next().asInstanceOf[UnsafeRow]
+            nextGroup = grouping(nextRow)
+          } else {
+            nextRow = null
+            nextGroup = null
+          }
+        }
+        fetchNextRow()
+
+        // Manage the current partition.
+        val buffer: ExternalAppendOnlyUnsafeRowArray =
+          new ExternalAppendOnlyUnsafeRowArray(inMemoryThreshold, spillThreshold)
+        var bufferIterator: Iterator[UnsafeRow] = _
+
+        val indexRow = new SpecificInternalRow(Array.fill(numBoundIndices)(IntegerType))
+
+        val frames = factories.map(_(indexRow))
+
+        private[this] def fetchNextPartition() {
+          // Collect all the rows in the current partition.
+          // Before we start to fetch new input rows, make a copy of nextGroup.
+          val currentGroup = nextGroup.copy()
+
+          // clear last partition
+          buffer.clear()
+
+          while (nextRowAvailable && nextGroup == currentGroup) {
```

Comment:
I guess we have to use `GenerateOrdering.compare()` to support complex types as well as `GroupedIterator` does.

Reply:
Good catch. @ueshin Do you mind double checking what I am doing now is correct?

Reply:
oh, we needed to fix the original `WindowExec` as well, didn't we.
How about fixing it in a separate PR first to see feedbacks from other reviewers who are familiar with window functions, and following the PR?

Reply:
I did fix the original `WindowExec` as well but of course we can fix it in a separate PR. I will revert my latest change.

Reply:
Thanks! Could you submit the PR to fix `WindowExec` please?

Reply:
Sure. Here https://github.com/apache/spark/pull/23279

Reply:
Why is this needed? You are comparing unsafe rows, so equality should work.

Reply:
@hvanhovell Thanks for chiming in. I am not very familiar with unsafe row comparison and was just following @ueshin suggestion. If this is not needed. I can close #23279 and leave it as is.

Reply:
Oh, I see. yeah, it should work. I missed that they are `UnsafeRow`s. Thanks!
@icexelloss Could you close #23279 and leave it as is? I am sorry for taking your time.

Reply:
@ueshin No worries! I closed #23279 and thank you @hvanhovell